dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/NaiveDevirtualisation.java

```diff
@@ -131,61 +136,108 @@ private void devirtualise(Function function, Map<Function, IValue> func2AddressM
+                caseLabels.add(caseLabel);
+                caseJumps.add(caseJump);
+            }
+
+            final Event noMatch = EventFactory.newAssert(expressions.makeFalse(), "Invalid function pointer");
+            final Label endLabel = EventFactory.newLabel(String.format("__Ldevirt_end#%s", devirtCounter));
+
+            final List<Event> callReplacement = new ArrayList<>();
+            callReplacement.add(EventFactory.newStringAnnotation("=== Devirtualized call ==="));
+            callReplacement.addAll(caseJumps);
+            callReplacement.add(noMatch);
+            for (int i = 0; i < caseLabels.size(); i++) {
+                callReplacement.add(caseLabels.get(i));
+                callReplacement.add(devirtualiseCall(call, possibleTargets.get(i)));
+                callReplacement.add(EventFactory.newGoto(endLabel));
+            }
+            callReplacement.add(endLabel);
+            callReplacement.add(EventFactory.newStringAnnotation("=== End of devirtualized call ==="));
+
+            call.replaceBy(callReplacement);
+            callReplacement.forEach(e -> e.copyAllMetadataFrom(call));
 
-                call.replaceBy(callReplacement);
-                callReplacement.forEach(e -> e.copyAllMetadataFrom(call));
+            devirtCounter++;
+        }
+    }
+
+    private boolean needsDevirtualization(FunctionCall call) {
+        return !call.isDirectCall() ||
+                (call.getCalledFunction().getIntrinsicInfo() == Intrinsics.Info.P_THREAD_CREATE
+                && !(call.getArguments().get(2) instanceof Function));
+    }
+
+    private List<Function> getPossibleTargets(FunctionCall call, Map<Function, IValue> func2AddressMap) {
+        final List<Function> possibleTargets;
+        if (!call.isDirectCall()) {
+            possibleTargets = func2AddressMap.keySet().stream()
+                    .filter(f -> f.getFunctionType() == call.getCallType()).collect(Collectors.toList());
+        } else if (call.getCalledFunction().getIntrinsicInfo() == Intrinsics.Info.P_THREAD_CREATE) {
+            final TypeFactory types = TypeFactory.getInstance();
+            final Type ptrType = types.getPointerType();
+            final Type threadType = types.getFunctionType(ptrType, List.of(ptrType));
+            possibleTargets = func2AddressMap.keySet().stream()
+                    .filter(f -> f.getFunctionType() == threadType).collect(Collectors.toList());
+        } else {
+            possibleTargets = List.of();
+            assert false;
+        }
 
-                devirtCounter++;
+        return possibleTargets;
+    }
+
+    private Expression getFunctionPointer(FunctionCall call) {
+        if (!call.isDirectCall()) {
+            return call.getCallTarget();
+        } else if (call.getCalledFunction().getIntrinsicInfo() == Intrinsics.Info.P_THREAD_CREATE) {
+            return call.getArguments().get(2);
+        }
+        assert false;
+        return null;
```

Comment:
Same as comment above.

The fact that you need to add a `return null` to make the compiler not complain about a missing return type already says we should not be using `assert false`

Reply:
It sucks that `assert false` still requires a return afterwards. I guess this is because `asserts` can get compiled away, leading to ill-formed code.